dev/src/expression.ts

```diff
@@ -0,0 +1,1425 @@
+    return {
+      arrayValue: {
+        values: this.exprs.map(p => serializer.encodeValue(p)!),
+      },
+    };
+  }
+}
+
+export class Field extends Expr implements Selectable {
+  exprType: ExprType = 'Field';
+  selectable = true as const;
+
+  private constructor(
+    private fieldPath: firestore.FieldPath,
+    private pipeline: Pipeline | null = null
+  ) {
+    super();
+  }
+
+  static of(name: string): Field;
+  static of(path: firestore.FieldPath): Field;
+  static of(nameOrPath: string | firestore.FieldPath): Field;
+  static of(pipeline: Pipeline, name: string): Field;
+  static of(
+    pipelineOrName: Pipeline | string | firestore.FieldPath,
+    name?: string
+  ): Field {
+    if (typeof pipelineOrName === 'string') {
+      if (FieldPath.documentId().formattedName === pipelineOrName) {
+        return new Field(new FieldPath('__path__'));
+      }
+
+      return new Field(FieldPath.fromArgument(pipelineOrName));
+    } else if (pipelineOrName instanceof FieldPath) {
+      if (FieldPath.documentId().isEqual(pipelineOrName)) {
+        return new Field(new FieldPath('__path__'));
+      }
+      return new Field(pipelineOrName);
+    } else {
+      return new Field(
+        FieldPath.fromArgument(name!),
+        pipelineOrName as Pipeline
+      );
+    }
+  }
+
+  fieldName(): string {
+    return (this.fieldPath as FieldPath).formattedName;
+  }
+
+  _toProto(serializer: Serializer): api.IValue {
+    return {
+      fieldReferenceValue: (this.fieldPath as FieldPath).formattedName,
+    };
+  }
+}
+
+export class Fields extends Expr implements Selectable {
+  exprType: ExprType = 'Field';
+  selectable = true as const;
+
+  private constructor(private fields: Field[]) {
+    super();
+  }
+
+  static of(name: string, ...others: string[]): Fields {
+    return new Fields([Field.of(name), ...others.map(Field.of)]);
+  }
+
+  static ofAll(): Fields {
+    return new Fields([]);
+  }
+
+  fieldList(): Field[] {
+    return this.fields.map(f => f);
+  }
+
+  _toProto(serializer: Serializer): api.IValue {
+    return {
+      arrayValue: {
+        values: this.fields.map(f => f._toProto(serializer)),
+      },
+    };
+  }
+}
+
+export class Constant extends Expr {
+  exprType: ExprType = 'Constant';
+
+  private constructor(private value: any) {
+    super();
+  }
+
+  static of(value: number): Constant;
+  static of(value: string): Constant;
+  static of(value: boolean): Constant;
+  static of(value: null): Constant;
+  static of(value: undefined): Constant;
+  static of(value: GeoPoint): Constant;
+  static of(value: Timestamp): Constant;
+  static of(value: Date): Constant;
+  static of(value: Uint8Array): Constant;
+  static of(value: DocumentReference): Constant;
+  static of(value: api.IValue): Constant;
+  static of(value: Array<any>): Constant;
+  static of(value: Map<string, any>): Constant;
+  static of(value: VectorValue): Constant;
+  static of(value: api.IValue): Constant;
```

Comment:
If you want to make this more extensible, you could also add an interface

```
interface FirestoreProtoSerializable {
  toProto(): api.IValue;
}
```

and have a `static of(value: FirestoreProtoSerializable): Constant`